client/src/app/users/user.service.spec.ts

```diff
@@ -59,135 +60,249 @@ describe('User service: ', () => {
-        && request.params.has('role') && request.params.has('company') && request.params.has('age')
-    );
-
-    // Check that the request made to that URL was a GET request.
-    expect(req.request.method).toEqual('GET');
+  describe('When getUsers() is called with parameters, it correctly forms the HTTP request (Javalin/Server filtering)', () => {
+    /*
+    * We really don't care what `getUsers()` returns in the cases
+    * where the filtering is happening on the server. Since all the
+    * filtering is happening on the server, `getUsers()` is really
+    * just a "pass through" that returns whatever it receives, without
+    * any "post processing" or manipulation. So the tests in this
+    * `describe` block all confirm that the HTTP request is properly formed
+    * and sent out in the world, but don't _really_ care about
+    * what `getUsers()` returns as long as it's what the HTTP
+    * request returns.
+    *
+    * So in each of these tests, we'll keep it simple and have
+    * the (mocked) HTTP request return the entire list `testUsers`
+    * even though in "real life" we would expect the server to
+    * return return a filtered subset of the users.
+    */
+
+    it('correctly calls api/users with filter parameter \'admin\'', () => {
+        const mockedMethod = spyOn(httpClient, 'get').and.returnValue(of(testUsers));
+
+        userService.getUsers({ role: 'admin' }).subscribe((users: User[]) => {
+          // The array of `User`s returned by `getUsers()` should be
+          // the array `testUsers`. This is "weird" because we'd truly be expecting
+          // the server to return just `admin` users, but as mentioned above, we're
+          // not trying to get the server here.
+          expect(users)
+            .withContext('expected users')
+            .toEqual(testUsers);
+          expect(mockedMethod)
+            .withContext('one call')
+            .toHaveBeenCalledTimes(1);
+          // The mocked method should have been called with two arguments:
+          //   * the appropriate URL ('/api/users' defined in the `UserService`)
+          //   * An options object containing an `HttpParams` with the `role`:`admin`
+          //     key-value pair.
+          expect(mockedMethod)
+            .withContext('talks to the correct endpoint')
+            .toHaveBeenCalledWith(userService.userUrl, { params: new HttpParams().set('role', 'admin') });
+        });
+    });
 
-    // Check that the role, company, and age parameters are correct
-    expect(req.request.params.get('role')).toEqual('editor');
-    expect(req.request.params.get('company')).toEqual('IBM');
-    expect(req.request.params.get('age')).toEqual('37');
+    it('correctly calls api/users with filter parameter \'age\'', () => {
+      const mockedMethod = spyOn(httpClient, 'get').and.returnValue(of(testUsers));
+
+      userService.getUsers({ age: 25 }).subscribe((users: User[]) => {
+        expect(users)
```

Comment:
Drop this as well?

Reply:
Do we drop the three lines so this `expect` goes away? (And, to be clear, it goes away since are is filtered on the server anyway.)

Reply:
It's the `expect` that I'm suggesting we drop (unless we want to keep the tests that check that `getUsers()` returns what the HTTP request returns.